### PR TITLE
Graceful shutdown of consumers

### PR DESCRIPTION
**Pull request by aavarghese**

### CODE REVIEW

**Comment by aavarghese** [comment]
Two issues so far:
1. Deletion of Redisstreamsource CR doesn't delete the adapter statefulset/pods even though owner references are set correctly.
2. Deletion of adapter pod itself sends the SIGTERM and prints out
```{"level":"info","ts":"2020-10-01T22:20:12.611Z","logger":"redis-stream-source","caller":"adapter/adapter.go:125","msg":"Listening for messages","stream":"mystream","consumerName":"redissource-mystream-1234-0-4"}
{"level":"info","ts":"2020-10-01T22:20:12.611Z","logger":"redis-stream-source","caller":"adapter/adapter.go:125","msg":"Listening for messages","stream":"mystream","consumerName":"redissource-mystream-1234-0-2"}
{"level":"info","ts":"2020-10-01T22:23:50.039Z","logger":"redis-stream-source","caller":"adapter/adapter.go:176","msg":"Consumer acknowledged the message","stream":"mystream","consumerName":"redissource-mystream-1234-0-0"}
{"level":"info","ts":"2020-10-01T22:23:55.568Z","logger":"redis-stream-source","caller":"adapter/adapter.go:185","msg":"Quit signal received, gracefully shutdown all consumers","stream":"mystream"}```

I don't see the ```a.logger.Info("Shutdown consumer", zap.String("consumerName", consumerName))``` from each consumer go routine that I was also expecting on SIGTERM!! How is it not falling into the ctx.Done case?

File: source/pkg/adapter/adapter.go [hunk 1]
```diff
@@ -59,25 +64,33 @@ func NewAdapter(ctx context.Context, processed adapter.EnvConfigAccessor, ceClie
 }
 
 func (a *Adapter) Start(ctx context.Context) error {
+	quitChannel := make(chan os.Signal, 1)
+	signal.Notify(quitChannel, syscall.SIGINT, syscall.SIGTERM)
```
Comment (lionelvillard) [inline source/pkg/adapter/adapter.go:68]: this is not needed: `ctx` is cancelled (ie. ctx->Done() is closed) when receiving SIGINT and SIGTERM signals

File: source/pkg/adapter/adapter.go [hunk 2]
```diff
@@ -92,30 +105,46 @@ func (a *Adapter) Start(ctx context.Context) error {
...
+					if err != nil {
+						a.logger.Error("Cannot delete consumer", zap.Error(err))
+					}
 
+					conn.Close()
+					return
+				default:
+				}
+				//XREAD below reads all the pending messages when xreadID=="0" and new messages when xreadID==">"
+				reply, err := conn.Do("XREADGROUP", "GROUP", groupName, consumerName, "COUNT", 1, "BLOCK", 0, "STREAMS", streamName, xreadID)
```
Comment (lionelvillard) [inline source/pkg/adapter/adapter.go:147]: this call is blocking so you may want to add a timeout (maybe 5 seconds?)

File: source/pkg/adapter/adapter.go [hunk 3]
```diff
@@ -92,30 +105,46 @@ func (a *Adapter) Start(ctx context.Context) error {
...
 
 			for {
-				//XREAD below reads all the pending messages when xreadID=="0" and new messages when xreadID==">"
-				reply, err := conn.Do("XREADGROUP", "GROUP", groupName, consumerName, "COUNT", 1, "BLOCK", 0, "STREAMS", a.config.Stream, xreadID)
+				//TODO: Check if statefulset's adapter pod is killed, and do not read more messages (to finish up work)
+				select {
+				case <-shutdownChannel:
+					a.logger.Info("Shutdown consumer", zap.String("consumerName", consumerName))
+
+					_, err := conn.Do("XGROUP", "DELCONSUMER", streamName, groupName, consumerName)
```
Comment (lionelvillard) [inline source/pkg/adapter/adapter.go:137]: what happens to the pending messages? Maybe there are none?

Comment (aavarghese) [reply source/pkg/adapter/adapter.go:159]: I don't think we can be sure of that. So better to check for pending maybe

**Comment by aavarghese** [comment]
@lionelvillard it all works now. Both deleting the pod and deleting the redis source (eventually deleting the pod) gracefully shuts down the consumer groups after processing pending messages.

Will try to see if I can refactor common code in both the select cases.

File: source/pkg/adapter/adapter.go [hunk 4]
```diff
@@ -129,66 +121,93 @@ func (a *Adapter) Start(ctx context.Context) error {
...
+							a.logger.Error("Failed to send cloudevent", zap.Any("result", result))
+						}
+
+						_, err = conn.Do("XACK", streamName, groupName, event.ID())
+						if err != nil {
+							a.logger.Error("Cannot ack message", zap.Error(err))
+							morePendingMsgs = true
+							continue
+						}
+						a.logger.Info("Consumer acknowledged the message", zap.String("consumerName", consumerName))
```
Comment (lionelvillard) [inline source/pkg/adapter/adapter.go:156]: This code looks very similar to the code below. Just checking `xreadID == ">"` should be enough to exit the goroutine. WDYT?

Comment (aavarghese) [reply source/pkg/adapter/adapter.go]: Yep thats enough :)

**Pull request merged**